model-evaluation/src/main/java/ai/vespa/models/handler/ModelsEvaluationHandler.java

```diff
@@ -1,47 +1,209 @@
+        // The following logic follows from the spec, in that signature and
+        // output are optional if the model only has a single function.
+
+        if (path.segments() == 3) {
+            if (model.functions().size() > 1) {
+                return listModelDetails(request, modelName.get());
+            }
+            return listTypeDetails(request, modelName.get());
+        }
+
+        if (path.segments() == 4) {
+            if ( ! path.segment(3).get().equalsIgnoreCase(EVALUATE)) {
+                return listTypeDetails(request, modelName.get(), path.segment(3).get());
+            }
+            if (model.functions().stream().anyMatch(f -> f.getName().equalsIgnoreCase(EVALUATE))) {
+                return listTypeDetails(request, modelName.get(), path.segment(3).get());  // model has a function "eval"
+            }
+            if (model.functions().size() <= 1) {
+                return evaluateModel(request, modelName.get());
+            }
+            return new ErrorResponse(400, "attempt to evaluate model without specifying function");
+        }
+
+        if (path.segments() == 5) {
+            if (path.segment(4).get().equalsIgnoreCase(EVALUATE)) {
+                return evaluateModel(request, modelName.get(), path.segment(3).get());
+            }
+        }
+
+        return new ErrorResponse(400, "unrecognized request");
+    }
+
+    private HttpResponse listAllModels(HttpRequest request) {
+        Slime slime = new Slime();
+        Cursor root = slime.setObject();
+        for (String modelName: modelsEvaluator.models().keySet()) {
+            root.setString(modelName, baseUrl(request) + modelName);
+        }
+        return new Response(200, com.yahoo.slime.JsonFormat.toJsonBytes(slime));
+    }
+
+    private HttpResponse listModelDetails(HttpRequest request, String modelName) {
+        Model model = modelsEvaluator.models().get(modelName);
+        Slime slime = new Slime();
+        Cursor root = slime.setObject();
+        for (ExpressionFunction func : model.functions()) {
+            root.setString(func.getName(), baseUrl(request) + modelName + "/" + func.getName());
+        }
+        return new Response(200, com.yahoo.slime.JsonFormat.toJsonBytes(slime));
+    }
+
+    private HttpResponse listTypeDetails(HttpRequest request, String modelName) {
+        return listTypeDetails(request, modelsEvaluator.evaluatorOf(modelName));
+    }
+
+    private HttpResponse listTypeDetails(HttpRequest request, String modelName, String signatureAndOutput) {
+        return listTypeDetails(request, modelsEvaluator.evaluatorOf(modelName, signatureAndOutput));
+    }
+
+    private HttpResponse listTypeDetails(HttpRequest request, FunctionEvaluator evaluator) {
+        Slime slime = new Slime();
+        Cursor root = slime.setObject();
+        Cursor bindings = root.setArray("bindings");
+        for (String bindingName : evaluator.context().names()) {
+            if (bindingName.startsWith("constant(")) {
```

Comment:
Ok, but I think we need another way to do this which only exposes the external bindings.